### ISSUE TITLE
Sidekiq::Limiter::Middleware calls #name on error instance

### ISSUE THREAD

**Issue by jbielick** [mperham/sidekiq#3604]
Ruby version: `ruby 2.2.7p470 (2017-03-28 revision 58194)`

Sidekiq / Pro / Enterprise version(s):

```
sidekiq (4.2.9)
sidekiq-ent (1.2.0)
sidekiq-pro (3.4.1)
```

Hi Mike—

We're happily using the `Limiter` functionality in `sidekiq-ent`, but have come across an issue adding a custom error to the config. We'd like for the limiter to catch `RestClient::TooManyRequests` errors and reschedule the job, however, when these errors occur, the limiter middleware attempts to call `#name` on the error instance. `*Sidekiq::Limiter.errors` types are being rescued and assigned to `x` and later `x.name` is called. I didn't expect an error class to implement `name`. Is this looking for the error class name, job class name? Please advise.

Thanks so much!

### Initializer

```ruby
# ...

Sidekiq::Limiter.configure do |config|
  config.errors.concat(
    [
      RestClient::TooManyRequests
    ]
  )
end

# ...
```

Forgive me if I should share more of the initializer. It's lengthy and I didn't think other details were relevant.

#### Error and Backtrace

```
NoMethodError: undefined method `name' for #<RestClient::TooManyRequests:0x...>
50             raise x
51           else
52             worker.logger.info { "#{x.name}: over rate limit, rescheduling for later" }
53             reschedule(worker, x, job, nil)
54           end
[GEM_ROOT]/gems/sidekiq-ent-1.2.0/lib/sidekiq-ent/limiter/middleware.rb:52 :in `block in call`
/usr/local/lib/ruby/2.2.0/logger.rb:372 :in `add`
/usr/local/lib/ruby/2.2.0/logger.rb:434 :in `info`
[GEM_ROOT]/gems/sidekiq-ent-1.2.0/lib/sidekiq-ent/limiter/middleware.rb:52 :in `rescue in call`
[GEM_ROOT]/gems/sidekiq-ent-1.2.0/lib/sidekiq-ent/limiter/middleware.rb:37 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/server/retry_jobs.rb:74 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/server/logging.rb:11 :in `block in call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/logging.rb:32 :in `with_context`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/server/logging.rb:7 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:133 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:133 :in `invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:132 :in `block (2 levels) in process`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:174 :in `stats`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:131 :in `block in process`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq.rb:35 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq.rb:35 :in `block in <module:Sidekiq>`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:126 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:126 :in `process`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:82 :in `process_one`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:70 :in `run`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/util.rb:17 :in `watchdog`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/util.rb:26 :in `block in safe_thread`
```

Since this exception happened inside a rescue, here's the outer exception:

```
RestClient::TooManyRequests: 429 Too Many Requests
[GEM_ROOT]/gems/rest-client-1.8.0/lib/restclient/abstract_response.rb:74 :in `return!`
[GEM_ROOT]/gems/rest-client-1.8.0/lib/restclient/request.rb:421 :in `block in transmit`
/usr/local/lib/ruby/2.2.0/net/http.rb:853 :in `start`
[GEM_ROOT]/gems/rest-client-1.8.0/lib/restclient/request.rb:413 :in `transmit`
[GEM_ROOT]/gems/rest-client-1.8.0/lib/restclient/request.rb:176 :in `execute`
[GEM_ROOT]/gems/rest-client-1.8.0/lib/restclient/request.rb:41 :in `execute`
[GEM_ROOT]/gems/rest-client-1.8.0/lib/restclient/resource.rb:67 :in `post`
[GEM_ROOT]/bundler/gems/{some api lib}/client.rb:145 :in `req`
[GEM_ROOT]/bundler/gems/{some api lib}/entities.rb:139 :in `create`
[GEM_ROOT]/bundler/gems/{some api lib}/client.rb:314 :in `create`
[PROJECT_ROOT]/app/models/{some class}.rb:323 :in `{some method}`
[PROJECT_ROOT]/app/models/{some class}.rb:206 :in `{some method}`
[PROJECT_ROOT]/app/models/{some class}.rb:783 :in `block in create_remote_campaign`
[PROJECT_ROOT]/app/models/concerns/active_record_semaphore.rb:95 :in `in_semaphore`
[PROJECT_ROOT]/app/models/{some class}.rb:777 :in `create_remote_campaign`
[PROJECT_ROOT]/app/models/{some class}.rb:631 :in `block in create_remote_campaign`
[PROJECT_ROOT]/app/models/concerns/active_record_semaphore.rb:95 :in `in_semaphore`
[PROJECT_ROOT]/app/models/{some class}.rb:630 :in `create_remote_campaign`
[PROJECT_ROOT]/lib/workers/{some class}/worker.rb:43 :in `block (2 levels) in perform`
[GEM_ROOT]/gems/statsd-ruby-1.2.1/lib/statsd.rb:232 :in `time`
[GEM_ROOT]/gems/hosted_graphite-0.0.8/lib/hosted_graphite/statsd.rb:55 :in `time`
[PROJECT_ROOT]/lib/workers/{some class}/worker.rb:42 :in `block in perform`
[GEM_ROOT]/gems/sidekiq-ent-1.2.0/lib/sidekiq-ent/limiter/bucket.rb:72 :in `block in within_limit`
[GEM_ROOT]/gems/sidekiq-ent-1.2.0/lib/sidekiq-ent/limiter/bucket.rb:61 :in `loop`
[GEM_ROOT]/gems/sidekiq-ent-1.2.0/lib/sidekiq-ent/limiter/bucket.rb:61 :in `within_limit`
[PROJECT_ROOT]/lib/workers/{some class}/worker.rb:28 :in `perform`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:158 :in `execute_job`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:138 :in `block (4 levels) in process`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq.rb:36 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq.rb:36 :in `block in <module:Sidekiq>`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:133 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:133 :in `block (3 levels) in process`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:128 :in `block in invoke`
[GEM_ROOT]/gems/statsd-ruby-1.2.1/lib/statsd.rb:232 :in `time`
[GEM_ROOT]/gems/sidekiq-pro-3.4.1/lib/sidekiq/middleware/server/statsd.rb:30 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[PROJECT_ROOT]/lib/sidekiq_middleware/scheduled_perform_tracker.rb:5 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-throttler-0.5.1/lib/sidekiq/throttler.rb:36 :in `block in call`
[GEM_ROOT]/gems/sidekiq-throttler-0.5.1/lib/sidekiq/throttler/rate_limit.rb:149 :in `call`
[GEM_ROOT]/gems/sidekiq-throttler-0.5.1/lib/sidekiq/throttler/rate_limit.rb:149 :in `execute`
[GEM_ROOT]/gems/sidekiq-throttler-0.5.1/lib/sidekiq/throttler.rb:43 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/newrelic_rpm-3.9.9.275/lib/new_relic/agent/instrumentation/sidekiq.rb:33 :in `block in call`
[GEM_ROOT]/gems/newrelic_rpm-3.9.9.275/lib/new_relic/agent/instrumentation/controller_instrumentation.rb:366 :in `perform_action_with_newrelic_trace`
[GEM_ROOT]/gems/newrelic_rpm-3.9.9.275/lib/new_relic/agent/instrumentation/sidekiq.rb:29 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/server/active_record.rb:6 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-pro-3.4.1/lib/sidekiq/batch/middleware.rb:33 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-ent-1.2.0/lib/sidekiq-ent/unique.rb:103 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-ent-1.2.0/lib/sidekiq-ent/limiter/middleware.rb:38 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/server/retry_jobs.rb:74 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/server/logging.rb:11 :in `block in call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/logging.rb:32 :in `with_context`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/server/logging.rb:7 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:130 :in `block in invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:133 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/middleware/chain.rb:133 :in `invoke`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:132 :in `block (2 levels) in process`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:174 :in `stats`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:131 :in `block in process`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq.rb:35 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq.rb:35 :in `block in <module:Sidekiq>`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:126 :in `call`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:126 :in `process`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:82 :in `process_one`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/processor.rb:70 :in `run`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/util.rb:17 :in `watchdog`
[GEM_ROOT]/gems/sidekiq-4.2.9/lib/sidekiq/util.rb:26 :in `block in safe_thread`
```

**Comment by mperham** [comment]
Yep, this is fixed in 1.6.1, which was released seconds ago.

**Comment by jbielick** [comment]
Will upgrade. Thanks!
